src/topology.jl

```diff
@@ -190,3 +190,189 @@ function Cart_sub(comm::Comm, remain_dims)
+
+Return the number of in and out edges for the calling processes in a distributed graph topology and a flag indicating whether the distributed graph is weighted.
+
+# External links
+$(_doc_external("MPI_Dist_graph_neighbors_count"))
+"""
+function Dist_graph_neighbors_count(graph_comm::Comm)
+    indegree = Ref{Cint}()
+    outdegree = Ref{Cint}()
+    weighted = Ref{Cint}()
+    # int MPI_Dist_graph_neighbors_count(MPI_Comm comm, int *indegree, int *outdegree, int *weighted)
+    @mpichk ccall((:MPI_Dist_graph_neighbors_count, libmpi), Cint,
+        (MPI_Comm, Ptr{Cint}, Ptr{Cint}, Ptr{Cint}),
+        graph_comm,indegree,outdegree,weighted)
+    (indegree[], outdegree[], weighted[])
+end
+
+"""
+    Dist_graph_neighbors!(graph_comm::Comm, sources::Vector{Cint}, source_weights::Vector{Cint}, destinations::Vector{Cint}, destination_weights::Vector{Cint})
+
+Return the neighbors of the calling process in a distributed graph topology without edge weights.
+
+# External links
+$(_doc_external("MPI_Dist_graph_neighbors"))
+"""
+function Dist_graph_neighbors!(graph_comm::Comm, sources::Vector{Cint}, destinations::Vector{Cint})
+    source_weights = Array{Cint}(undef,0)
+    destination_weights = Array{Cint}(undef,0)
+    # int MPI_Dist_graph_neighbors(MPI_Comm comm,
+    #       int maxindegree, int sources[], int sourceweights[],
+    #       int maxoutdegree, int destinations[], int destweights[])
+    @mpichk ccall((:MPI_Dist_graph_neighbors, libmpi), Cint,
+        (MPI_Comm, Cint, Ptr{Cint}, Ptr{Cint}, Cint, Ptr{Cint}, Ptr{Cint}),
+        graph_comm,length(sources),sources,source_weights,length(destinations),destinations,destination_weights)
+end
+
+"""
+    Dist_graph_neighbors!(graph_comm::Comm, sources::Vector{Cint}, source_weights::Vector{Cint}, destinations::Vector{Cint}, destination_weights::Vector{Cint})
+
+Return the neighbors and edge weights of the calling process in a distributed graph topology.
+
+# External links
+$(_doc_external("MPI_Dist_graph_neighbors"))
+"""
+function Dist_graph_neighbors!(graph_comm::Comm, sources::Vector{Cint}, source_weights::Vector{Cint}, destinations::Vector{Cint}, destination_weights::Vector{Cint})
+    # int MPI_Dist_graph_neighbors(MPI_Comm comm,
+    #       int maxindegree, int sources[], int sourceweights[],
+    #       int maxoutdegree, int destinations[], int destweights[])
+    @mpichk ccall((:Dist_graph_neighbors, libmpi), Cint,
+        (MPI_Comm, Cint, Ptr{Cint}, Ptr{Cint}, Cint, Ptr{Cint}, Ptr{Cint}),
+        graph_comm,length(sources),sources,source_weights,length(destinations),destinations,destination_weights)
+end
+
+"""
+    Neighbor_alltoall!(sendbuf::UBuffer, recvbuf::UBuffer, comm::Comm)
+
+Perform an all-to-all communication along the directed edges of the graph with fixed size messages.
+
+See also [`MPI.alltoall!`](@ref).
+
+# External links
+$(_doc_external("MPI_Neighbor_alltoall"))
+"""
+function Neighbor_alltoall!(sendbuf::UBuffer, recvbuf::UBuffer, graph_comm::Comm)
```

Comment:
Since I have put the neighbor topology primitives in `topology.jl` I decided to put the communication primitives in `collective.jl`. If you want I can also create two new files and put the new wrappers into them (i.e. `neighbor_topology.jl` and `neighbor_collective.jl`).